pkg/standalone-utils/test/LidoRelayer.test.ts

```diff
@@ -1,462 +1,950 @@
-          await expect(relayer.connect(sender).swap(singleSwap, funds, limit, deadline)).to.be.revertedWith(
-            'USER_DOESNT_ALLOW_RELAYER'
+        it('unwraps with chained references', async () => {
+          const amount = fp(1);
+          await setChainedReferenceContents(toChainedReference(0), amount);
+
+          const receipt = await (
+            await relayer.connect(sender).multicall([encodeUnwrap(tokenSender, tokenRecipient, toChainedReference(0))])
+          ).wait();
+
+          expectTransferEvent(
+            receipt,
+            {
+              from: relayer.address,
+              to: ZERO_ADDRESS,
+              value: amount,
+            },
+            wstETH
           );
         });
-      });
+      }
     });
-  });
 
-  describe('batchSwap', () => {
-    let swaps: BatchSwapStep[];
-    let limits: BigNumberish[];
-    let assets: string[];
-    let funds: FundManagement;
-    const deadline = MAX_UINT256;
-
-    sharedBeforeEach('build fund management', async () => {
-      funds = {
-        sender: sender.address,
-        fromInternalBalance: false,
-        recipient: sender.address,
-        toInternalBalance: false,
-      };
-    });
+    describe('stakeETH', () => {
+      let tokenRecipient: Account;
 
-    context('when the relayer is authorized', () => {
-      sharedBeforeEach('allow relayer', async () => {
-        const manageUserBalanceAction = await actionId(vault.instance, 'manageUserBalance');
-        const batchSwapAction = await actionId(vault.instance, 'batchSwap');
+      context('sender = user, recipient = relayer', () => {
+        beforeEach(() => {
+          tokenRecipient = relayer;
+        });
+        testStake();
+      });
 
-        await vault.authorizer
-          ?.connect(admin)
-          .grantRolesGlobally([manageUserBalanceAction, batchSwapAction], relayer.address);
+      context('sender = relayer, recipient = relayer', () => {
+        beforeEach(() => {
+          tokenRecipient = relayer;
+        });
+        testStake();
       });
 
-      context('when the user did allow the relayer', () => {
-        sharedBeforeEach('allow relayer', async () => {
-          await vault.instance.connect(sender).setRelayerApproval(sender.address, relayer.address, true);
-        });
-
-        describe('swap using stETH as an input', () => {
-          sharedBeforeEach('build swap request', async () => {
-            swaps = [
-              {
-                poolId: basePoolId,
-                assetInIndex: basePoolTokens.findIndexBySymbol('wstETH'),
-                assetOutIndex: basePoolTokens.findIndexBySymbol('WETH'),
-                amount: fp(1),
-                userData: '0x',
-              },
-            ];
-
-            assets = basePoolTokens.addresses;
-            limits = basePoolTokens.map((token) => (token.symbol === 'wstETH' ? fp(1) : 0));
-          });
+      context('sender = relayer, recipient = sender', () => {
```

Comment:
This seems incorrect: the sender of stake is always the user (who sends eth to the relayer): we should only have two options, depending on whether the recipient is the relayer or the user.